Kotlin conversion

Fabi2607 opened the pull request:
Complete conversion of the code base to Kotlin, including some code cleanups.

Fixes #60

theZorro266 commented on app/src/main/java/com/example/paetz/yacguide/utils/NetworkUtils.kt:
```diff
@@ -0,0 +1,37 @@
+
+object NetworkUtils {
+
+    // I couldn't find a proper java library, so let's escape it on our own
```
https://fckaf.de/qw8

Fabi2607 replied:
Replaced with UrlEncoder

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/utils/WidgetUtils.kt:
```diff
@@ -0,0 +1,101 @@
+        return lineView
+    }
+
+    private fun createTextView(context: Context,
```
Does not rely comply with any style guides and I only now it from python code bases...

paetz replied:
At least I would like to have some style guide line, no matter what. Since most of the codebase uses this style by now, I would stick to it.

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/AscendActivity.kt:
```diff
@@ -0,0 +1,168 @@
+        if (_partnerIds == null) {
+            _partnerIds = if (_ascend == null) ArrayList() else _ascend!!.partnerIds
+        }
+        _day = 0
```
True, this is an artifact of auto conversion since Kotlin doesn't know about "onCreate" on a language level

Fabi2607 commented:
Most of the comments are caused by the auto conversion. I mostly cleaned stuff up that was obviously not working. I did not clean up stuff that could have been written in a nicer way. I already started going through the code, and wanted to submit them afterwards in small reviews.. But I want to avoid having multiple PRs open that depend on each other.

Should I add all the work I started doing to this PR and do you want to keep it open until then?

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/database/TourbookExporter.kt:
```diff
@@ -0,0 +1,126 @@
+        val file = File(filePath)
+        val jsonString: String
+        try {
+            jsonString = FileInputStream(file).bufferedReader().readText()
```
Yes, the documentation states UTF-8 as default

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/database/TourbookExporter.kt:
```diff
@@ -0,0 +1,126 @@
+            db.ascendDao().insert(ascend)
+            val route = db.routeDao().getRoute(routeId)
+            if (route != null) {
+                db.routeDao().updateAscendCount(db.routeDao().getAscendCount(routeId) + 1, routeId)
```
Was removed by accident

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/network/JSONWebParser.kt:
```diff
@@ -0,0 +1,50 @@
+abstract class JSONWebParser(protected var db: AppDatabase, protected var listener: UpdateListener) : NetworkListener {
+    protected val baseUrl = "http://db-sandsteinklettern.gipfelbuch.de/"
+    protected var networkRequests: LinkedList<NetworkRequest> = LinkedList()
+    private var success: Boolean = true
```
What I don't like about the leading underscores is, that it requires refactoring of a huge amount of code, just to change the visibility of one variable (even from private to protected), while not adding any value (it is already clearly stated that the variable is private, the compiler only allows it to be used within the class, where I can easily see it is private. Outside I can not use the variable anyway. I clearly get why it is done e.g. in Python, where you don't have visibility modifiers, but I never encountered it in a Java/Kotlin codebase.

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/network/NetworkTask.kt:
```diff
@@ -0,0 +1,32 @@
+            System.err.println(e)
+            null
+        } finally {
+            connection.disconnect()
```
Connection can't be null at this point

paetz commented:
I don't get your point. Sequence should be PR -> Review -> Review changes -> Merge. If you start new stuff during review that depends on the open PR, branch off from there and rebase after merging the PR.

Fabi2607 commented:
I tried to actively avoid making too many changes that go into a refactoring direction and potential change the behavior to be able to introduce them afterwards as independent easy to review changes.

I am happy to address it in this PR, in my experience it was just easier to manage big conversions by keeping the amount of additional changes small to get to a point where you can have easy to review changes again, instead of trying to review everything in one huge bunch of mixed stuff.

But if you prefer I will add all the changes to this PR.

paetz commented:
True, small PRs are always better. But in this case we will surely forget the review comments. You could surely create a follow-up issue for cleaning that stuff up, but this will lead to another large PR.
So I think, addressing the review comments here, is most effective to get rid of this huge PR asap.

Fabi2607 commented:
> True, small PRs are always better. But in this case we will surely forget the review comments. You could surely create a follow-up issue for cleaning that stuff up, but this will lead to another large PR.
> So I think, addressing the review comments here, is most effective to get rid of this huge PR asap.

Ok, then I will clean it up here, but I would opt to leave out stuff, that I already removed completely in my follow-up code. I will add a comment and reference to the PR that removes it for them. I will open that PR in parallel.

I will also keep #57 rebased on this branch at all time to track the current state of code I want to arrive at.

paetz commented:
Sounds good. To avoid merge conflicts, you don't need to add comments if you have addressed it in the follow-up PR.

Fabi2607 commented:
> Sounds good. To avoid merge conflicts, you don't need to add comments if you have addressed it in the follow-up PR.

I meant PR comments, not actual comments in code ;)

Fabi2607 commented on app/src/main/java/com/example/paetz/yacguide/network/JSONWebParser.kt:
```diff
@@ -0,0 +1,50 @@
+    override fun onNetworkTaskResolved(requestId: NetworkRequestType, result: String) {
+        if (success) {
+            try {
+                if (result.equals("null", ignoreCase = true)) {
```
`result` is never null here.

Fabi2607 commented:
I changed all of the comments. There are a few private functions remaining with wrong naming that will change anyway in the follow up PR.
